### PR TITLE
Introduce producer-type node

### PR DESCRIPTION
**Pull request by arvind**
A new `Produces` flag identifies nodes that generate new tuples. When a branch of the dataflow graph contains such nodes, Collectors are automatically inserted after them. The logic for this has also been extracted from `DataSource.pipeline` into `Graph.preprocess`. An updated `synchronize` method (which has been moved to `Graph` as well) traverses a dataflow branch and updates the previous values of tuples via Collectors.

### CODE REVIEW

**Comment by arvind** [comment]
This seemed the cleanest way to introduce this functionality, and is currently most applicable to transforms such as Aggregate, Cross, etc. Some options that I ruled out:
- _identifying these nodes as `Collectors` themselves and embedding calls to `Collector.prototype` within `Transform`_: not sufficiently generalized. More importantly, Collector-type nodes are used to connect external dependencies, and Transforms cannot be directly connected in this way.
- _embedding a `Collector` within a producing node_: Made batch transforms (which already contain a collector reference) more messy, and made it more difficult to effectively share collectors in `preprocess`.

File: src/Graph.js [hunk 1]
```diff
@@ -158,6 +160,43 @@ prototype.propagate = function(pulse, node, stamp) {
   }
 };
 
+// Process a new branch of the dataflow graph prior to connection:
+// (1) Insert new Collector nodes as needed. 
+// (2) Track + return mutation/routing status of the branch.
+prototype.preprocess = function(branch) {
+  var graph = this,
+      mutates = 0,
+      node, router, collector, collects;
```
Comment (jheer) [inline src/Graph.js:169]: The previous version of this method (in `DataSource`) initialized the `collector` variable to the `_inputNode`. I did this as it helped remove unnecessary collectors (e.g., consider a batch transform at the beginning of a pipeline). Here it is left `undefined` and we inject a new collector instead. Why the change?

Comment (arvind) [inline src/Graph.js:169]: Right, in a general call to this method, we will insert a collector. However, when called from `DataSource`, `_inputNode` is first unshifted onto `pipeline` and then passed to `preprocess`. The `_inputNode` is marked as a collector and, being the first node in the `branch`, should yield the same behaviour I think.
